Content.Server/Replays/ReplayEventSystem.cs

```diff
@@ -0,0 +1,160 @@
+        var bytes = Encoding.UTF8.GetBytes(events.ToString());
+        replayRecordingStopped.Writer.WriteBytes(replayRecordingStopped.Writer.BaseReplayPath / "events.yml", bytes);
+    }
+
+    /// <summary>
+    /// Records a replay event. This is the main way to record events in the replay system.
+    /// </summary>
+    /// <param name="replayEvent">The event to record</param>
+    /// <param name="source">Optional source that will be used for location data</param>
+    public override void RecordReplayEvent(ReplayEvent replayEvent, EntityUid? source = null)
+    {
+        if (!_replays.IsRecording || !_recordEvents)
+            return;
+
+        replayEvent.Time ??= _gameTicker.RoundDuration().TotalSeconds;
+
+        if (source.HasValue && replayEvent.Location == null)
+        {
+            replayEvent.Location = new LocationInformation
+            {
+                Position = _transformSystem.GetWorldPosition(source.Value),
+                NearestBeacon = _navMapSystem.GetNearestBeaconString(_transformSystem.GetMapCoordinates(source.Value)),
+            };
+
+            var map = _transformSystem.GetMap(source.Value);
+            if (map.HasValue)
+            {
+                replayEvent.Location.Map = EntityManager.GetComponent<MetaDataComponent>(map.Value).EntityName;
+            }
+        }
+
+        if (replayEvent.EventType == null || replayEvent.Severity == null)
+            throw new ArgumentException("Replay event must have a type and severity.");
+
+        Log.Verbose($"Recording replay event: {replayEvent.EventType}");
+        _replayEvents.Add(replayEvent);
+    }
+
+    /// <summary>
+    /// Gets the player info for a player entity.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">
+    /// Thrown if the entity is not a player.
+    /// </exception>
+    public override ReplayEventPlayer GetPlayerInfo(EntityUid player)
+    {
+        if (EntityManager.TryGetComponent<ActorComponent>(player, out var actorComponent))
+        {
+            return GetPlayerInfo(actorComponent.PlayerSession);
+        }
+
+        var stackTrace = new StackTrace();
+        Log.Warning($"Tried to get player info for entity {player}, but it's not a player entity. Stacktrace: {stackTrace}");
+        return new ReplayEventPlayer()
+        {
+            PlayerGuid = new NetUserId(Guid.Empty),
+            PlayerICName = EntityManager.GetComponent<MetaDataComponent>(player).EntityName, // Fallback, best we can do.
+            PlayerOOCName = "",
+            JobPrototypes = [],
+            AntagPrototypes = [],
+        };
+    }
+
+    /// <summary>
+    /// Generates a <see cref="ReplayEventPlayer"/> from a session for use in replay events.
+    /// </summary>
+    public override ReplayEventPlayer GetPlayerInfo(ICommonSession session)
+    {
+        var hasMind = _mindSystem.TryGetMind(session, out var mindId, out var mindComponent);
+
+        var playerIcName = "Unknown";
```

Comment:
This should default to null or some other value, let UIs figure out how to present it.